pydrofoil/real.py

```diff
@@ -0,0 +1,155 @@
+import sys
+from rpython.rlib.rbigint import rbigint, _divrem as bigint_divrem
+from rpython.rlib.rarithmetic import r_uint, intmask, string_to_int, ovfcheck, \
+        int_c_div, int_c_mod, r_ulonglong
+
+MAXINT = 2**63-1
+MININT = -2**63
+
+class Real(object):
+    def __init__(self, num, den, Normalize = True):
+        if Normalize:
+            self.num = num
+            self.den = den
+        else:
+            gcd_val = num.gcd(den)
+            self.num = num.div(gcd_val)
+            self.den = den.div(gcd_val)
+        
+
+    @staticmethod
+    def fromint(num, den=1):
+        assert not den==0, "denominator cannot be 0"
+        if num == MININT or den == MININT:
+            num = rbigint.fromint(num)
+            den = rbigint.fromint(den)
+            sign = num.sign * den.sign
+            num = num.abs().int_mul(sign)
+            den = den.abs()
+        elif num < MININT or num > MAXINT or den < MININT or den > MAXINT:
+            assert False, "input is out of range of INT"
+        else:
+            sign = num/abs(num)*den/abs(den)
+            num = rbigint.fromint(abs(num)*sign)
+            den = rbigint.fromint(abs(den))
+        return Real(num, den, False)
+    
+    @staticmethod
+    def fromstr(str):
+        from rpython.rlib.rstring import strip_spaces
+        s = strip_spaces(str)
+        decimalpos = s.find(".")
+        num = rbigint.fromstr(s[:decimalpos]+s[decimalpos+1:], 10) if decimalpos != -1 else rbigint.fromstr(s, 10)
+        den = rbigint.fromint(10).int_pow(len(s)-1 - decimalpos) if decimalpos != -1 else rbigint.fromint(1)
+        return Real(num, den, False)
+
+
+    def add(self, other):
+        g = self.den.gcd(other.den)
+        if g.int_eq(1):
+            num_new_1 = self.num.mul(other.den)
+            num_new_2 = other.num.mul(self.den)
+            den_new = self.den.mul(other.den)
+            num_new = num_new_1.add(num_new_2)
+            return Real(num_new, den_new)
+        s = self.den.div(g)
+        t = self.num.mul(other.den.div(g)).add(other.num.mul(s))
+        g2 = t.gcd(g)
+        if g2.int_eq(1):
+            return Real(t, s.mul(other.den))
+        return Real(t.div(g2), s.mul(other.den.div(g2)))
+
+    
+    def sub(self, other):
+        g = self.den.gcd(other.den)
+        if g.int_eq(1):
+            num_new_1 = self.num.mul(other.den)
+            num_new_2 = other.num.mul(self.den)
+            den_new = self.den.mul(other.den)
+            num_new = num_new_1.sub(num_new_2)
+            return Real(num_new, den_new)
+        s = self.den.div(g)
+        t = self.num.mul(other.den.div(g)).sub(other.num.mul(s))
+        g2 = t.gcd(g)
+        if g2.int_eq(1):
+            return Real(t, s.mul(other.den))
+        return Real(t.div(g2), s.mul(other.den.div(g2)))
+    
+    def mul(self, other):
+        g1 = self.num.gcd(other.den)
+        g2 = other.num.gcd(self.den)
```

Comment:
your `mul` and `div` are less efficient than the one in `fractions.py`: in `fractions.py` the second call to `gcd` already takes smaller input arguments if `g1 > 1`. This potentially makes the second `gcd` call more efficient, because the arguments are smaller.

I think it's probably best to follow the approach in `fractions.py` quite closely, the CPython developers thought about this much more carefully than we did ;-).

Reply:
ok thank you so much, I would change it

Reply:
sorry, I'm wondering whether there's a difference between putting g2 right after g1 or put it later after g1>1, cause the g2 is calculated from the second numerator and first denominator, which will not be updated if g1>1